### PR TITLE
remove extra class

### LINKED ISSUE #1194: Remove ExceptionAndPath
com.google.cloud.tools.opensource.dependencies.ExceptionAndPath is only used in one method in DependencyGraphBuilder where it is immediately unwrapped. We can go straight to the UnavilableArtifactException instead.

### PR DESCRIPTION
**Pull request by elharo**
fixes #1194 @suztomo

### CODE REVIEW

File: dependencies/src/main/java/com/google/cloud/tools/opensource/dependencies/UnresolvableArtifactProblem.java [hunk 1]
```diff
@@ -41,6 +42,18 @@ public UnresolvableArtifactProblem(Artifact artifact) {
   public UnresolvableArtifactProblem(List<DependencyNode> dependencyPath) {
     super(dependencyPath.get(dependencyPath.size() - 1).getArtifact(), dependencyPath);
   }
+  
+  /**
+   * Problem when Maven cannot resolve the artifact at the leaf of {@code dependencyPath} in a
+   * dependency tree.
+   * 
+   * @param artifact dependency that can't be found
+   * @param path to the unfound artifact, not including the unfound artifact
```
Comment (suztomo) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/dependencies/UnresolvableArtifactProblem.java]: `path` needs to include the unfound artifact; otherwise we cannot tell the scope / optional of the dependency.

Comment (elharo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/dependencies/UnresolvableArtifactProblem.java]: Let me double check, but I think that would be a behavioral change from the current code

Comment (elharo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/dependencies/UnresolvableArtifactProblem.java]: OK, I think you're right though the code is a little hard to follow. The parentNodes stack variable might be poorly named. In any case I can make this doc comment ambiguous enough to cover both cases, and this PR does not change the behavior

Comment (suztomo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/dependencies/UnresolvableArtifactProblem.java]: The parentNodes do not have the node being checked. ExceptionAndPath was combining them to make the dependency path for UnresolvableArtifactProblem.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/dependencies/ArtifactProblem.java [hunk 2]
```diff
@@ -35,7 +35,7 @@
 
   protected final ImmutableList<DependencyNode> dependencyPath;
 
-  protected ArtifactProblem(Artifact artifact, List<DependencyNode> dependencyPath) {
+  protected ArtifactProblem(Artifact artifact, Collection<DependencyNode> dependencyPath) {
```
Comment (suztomo) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/dependencies/ArtifactProblem.java]: As we care about the order of the dependency path, List is appropriate.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/dependencies/DependencyGraphBuilder.java [hunk 3]
```diff
@@ -340,16 +340,15 @@ private DependencyGraphResult levelOrder(
                 resolutionException.getResult().getArtifactResults()) {
               if (artifactResult.getArtifact() == null) {
                 DependencyNode failedDependencyNode = artifactResult.getRequest().getDependencyNode();
-                ExceptionAndPath failure =
-                    ExceptionAndPath.create(parentNodes, failedDependencyNode, resolutionException);
-                artifactProblems.add(new UnresolvableArtifactProblem(failure.getPath()));
+                artifactProblems.add(
+                    new UnresolvableArtifactProblem(failedDependencyNode.getArtifact(), parentNodes));
```
Comment (suztomo) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/dependencies/DependencyGraphBuilder.java]: Memo: the parentNodes.push(dependencyNode) happens at line 327 (before this line).

File: dependencies/src/main/java/com/google/cloud/tools/opensource/dependencies/UnresolvableArtifactProblem.java (hunk 1 above)
Comment (suztomo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/dependencies/UnresolvableArtifactProblem.java]: > The parentNodes do not have the node being checked.

This was wrong.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/dependencies/UnresolvableArtifactProblem.java [hunk 4]
```diff
@@ -41,6 +42,18 @@ public UnresolvableArtifactProblem(Artifact artifact) {
   public UnresolvableArtifactProblem(List<DependencyNode> dependencyPath) {
     super(dependencyPath.get(dependencyPath.size() - 1).getArtifact(), dependencyPath);
   }
+  
+  /**
+   * Problem when Maven cannot resolve the artifact at the leaf of {@code dependencyPath} in a
+   * dependency tree.
+   * 
+   * @param artifact dependency that can't be found
+   * @param path to the unfound artifact
```
Comment (suztomo) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/dependencies/UnresolvableArtifactProblem.java]:
```suggestion
   * @param path path to the unavailable artifact
```

Comment (suztomo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/dependencies/UnresolvableArtifactProblem.java]: Because the artifact is held by the last element of path, you don't need the constructor.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/dependencies/UnresolvableArtifactProblem.java [hunk 5]
```diff
@@ -41,6 +42,18 @@ public UnresolvableArtifactProblem(Artifact artifact) {
...
+   */
+  public UnresolvableArtifactProblem(Artifact artifact, Collection<DependencyNode> path) {
```
Comment (suztomo) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/dependencies/UnresolvableArtifactProblem.java]: Because the `artifact` is held by the last element of path, you don't need the constructor.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/dependencies/ArtifactProblem.java (hunk 2 above)
Comment (elharo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/dependencies/ArtifactProblem.java]: OK, fixed, though stacks being a subclass of List is a long-recognized problem in the Java Collections API since it violates the ADT for stack.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/dependencies/DependencyGraphBuilder.java (hunk 3 above)
Comment (elharo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/dependencies/DependencyGraphBuilder.java]: next PR

File: dependencies/src/main/java/com/google/cloud/tools/opensource/dependencies/UnresolvableArtifactProblem.java (hunk 5 above)
Comment (elharo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/dependencies/UnresolvableArtifactProblem.java]: done

File: dependencies/src/main/java/com/google/cloud/tools/opensource/dependencies/UnresolvableArtifactProblem.java [hunk 6]
```diff
@@ -37,6 +37,8 @@ public UnresolvableArtifactProblem(Artifact artifact) {
   /**
    * Problem when Maven cannot resolve the artifact at the leaf of {@code dependencyPath} in a
    * dependency tree.
+   * 
+   * @param path to the unavailable artifact
    */
```
Comment (suztomo) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/dependencies/UnresolvableArtifactProblem.java]: Incorrect param name.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/dependencies/DependencyGraphBuilder.java [hunk 7]
```diff
@@ -339,8 +339,11 @@ private DependencyGraphResult levelOrder(
             for (ArtifactResult artifactResult :
                 resolutionException.getResult().getArtifactResults()) {
               if (artifactResult.getArtifact() == null) {
-                artifactProblems.add(new UnresolvableArtifactProblem(parentNodes));
```
Comment (suztomo) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/dependencies/DependencyGraphBuilder.java]: Memo:

This fails:

image

File: dependencies/src/main/java/com/google/cloud/tools/opensource/dependencies/ExceptionAndPath.java [hunk 8]
```diff
@@ -1,67 +0,0 @@
...
-  }
-
-  static ExceptionAndPath create(
-      Iterable<DependencyNode> parentDependencyNodes,
-      DependencyNode dependencyNode,
-      RepositoryException repositoryException) {
-    return new ExceptionAndPath(
-        ImmutableList.<DependencyNode>builder()
-            .addAll(parentDependencyNodes)
-            .add(dependencyNode)
```
Comment (suztomo) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/dependencies/ExceptionAndPath.java]:
```
parentDependencyNodes:
  org.sonatype.sisu:sisu-guice:3.2.6 (compile)
   / cglib:cglib:3.1 (provided, optional)
   / ant:ant:1.6.2 (compile, optional)

dependencyNode: xerces:xerces-impl:jar:2.6.2 (compile?)
```
creating 
```
parentDependencyNodes:
  org.sonatype.sisu:sisu-guice:3.2.6 (compile)
   / cglib:cglib:3.1 (provided, optional)
   / ant:ant:1.6.2 (compile, optional) 
   / dependencyNode: xerces:xerces-impl:jar:2.6.2 (compile?)
```

**Comment by suztomo** [comment]
I got the cause of the strangeness:

In function levelOrder, it was visiting a dependency "ant:1.6.2" when DependencyResolutionException occured. Ant:1.6.2 is in Maven Central. However its dependencies "xerces-impl:2.6.2" and "xml-apis:2.6.2" are unavailable. That's why it received the exception.

image

File: dependencies/src/main/java/com/google/cloud/tools/opensource/dependencies/DependencyGraphBuilder.java [hunk 9]
```diff
@@ -339,17 +339,15 @@ private DependencyGraphResult levelOrder(
...
               if (artifactResult.getArtifact() == null) {
-                DependencyNode failedDependencyNode = artifactResult.getRequest().getDependencyNode();
-                ExceptionAndPath failure =
-                    ExceptionAndPath.create(parentNodes, failedDependencyNode, resolutionException);
-                artifactProblems.add(new UnresolvableArtifactProblem(failure.getPath()));
+                DependencyNode failedDependencyNode =
+                    artifactResult.getRequest().getDependencyNode();
+                artifactProblems.add(
+                    new UnresolvableArtifactProblem(
+                        failedDependencyNode.getArtifact(), parentNodes));
```
Comment (suztomo) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/dependencies/DependencyGraphBuilder.java]:
```suggestion
                        ImmutableList.builder().addAll(parentNodes).add(failedDependencyNode).build()));
```

In `ClassPathBuilderTest.testResolve_artifactProblems` failure, the last element of parentNodes does not match `failedDependencyNode.getArtifact()`, because "ant:1.6.2" is not the missing artifact.

image

**Comment by suztomo** [comment]
CircleCI says
```
Caused by: org.eclipse.aether.transfer.ArtifactTransferException: Could not transfer artifact com.google.guava:guava-bom:pom:28.2-android from/to central (https://repo.maven.apache.org/maven2): Authorization failed for https://repo.maven.apache.org/maven2/com/google/guava/guava-bom/28.2-android/guava-bom-28.2-android.pom 403 Forbidden
```

File: dependencies/src/main/java/com/google/cloud/tools/opensource/dependencies/DependencyGraphBuilder.java [hunk 10]
```diff
@@ -339,20 +339,22 @@ private DependencyGraphResult levelOrder(
...
-                ExceptionAndPath failure =
-                    ExceptionAndPath.create(parentNodes, failedDependencyNode, resolutionException);
-                artifactProblems.add(new UnresolvableArtifactProblem(failure.getPath()));
+                DependencyNode failedDependencyNode =
+                    artifactResult.getRequest().getDependencyNode();
+                
+                List<DependencyNode> fullPath = makeFullPath(parentNodes, failedDependencyNode);
+                
+                artifactProblems.add(
+                    new UnresolvableArtifactProblem(
```
Comment (suztomo) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/dependencies/DependencyGraphBuilder.java]:
```suggestion
                    new UnresolvableArtifactProblem(fullPath)
```

Then the new constructor is not needed.

File: dependencies/src/test/java/com/google/cloud/tools/opensource/dependencies/DependencyGraphBuilderTest.java [hunk 11]
```diff
@@ -201,8 +201,6 @@ public void testGetStaticLinkageCheckDependencyGraph_artifactProblems()
     assertEquals("xerces:xerces-impl:jar:2.6.2", firstProblem.getArtifact().toString());
 
     assertEquals(
-        "The problem should explain the dependency path to the unavailable artifact, "
-            + "including scope and optional flag",
```
Comment (suztomo) [inline dependencies/src/test/java/com/google/cloud/tools/opensource/dependencies/DependencyGraphBuilderTest.java]: Why is it being removed?

File: dependencies/src/main/java/com/google/cloud/tools/opensource/dependencies/DependencyGraphBuilder.java (hunk 10 above)
Comment (elharo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/dependencies/DependencyGraphBuilder.java]: removed

File: dependencies/src/test/java/com/google/cloud/tools/opensource/dependencies/DependencyGraphBuilderTest.java (hunk 11 above)
Comment (elharo) [reply dependencies/src/test/java/com/google/cloud/tools/opensource/dependencies/DependencyGraphBuilderTest.java]: This feels like more a comment about what the method does rather than why the test failed. When I ran the test this message pushed the actual diff out of the way.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/dependencies/DependencyGraphBuilder.java [hunk 12]
```diff
@@ -339,20 +339,22 @@ private DependencyGraphResult levelOrder(
...
+                        failedDependencyNode.getArtifact(), fullPath));
```
Comment (suztomo) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/dependencies/DependencyGraphBuilder.java]:
```suggestion
                        fullPath));
```

File: dependencies/src/test/java/com/google/cloud/tools/opensource/dependencies/DependencyGraphBuilderTest.java (hunk 11 above)
Comment (suztomo) [reply dependencies/src/test/java/com/google/cloud/tools/opensource/dependencies/DependencyGraphBuilderTest.java]: That's true. It was a comment I wanted to clarify the reason of this assertion.

File: dependencies/src/main/java/com/google/cloud/tools/opensource/dependencies/DependencyGraphBuilder.java (hunk 12 above)
Comment (elharo) [reply dependencies/src/main/java/com/google/cloud/tools/opensource/dependencies/DependencyGraphBuilder.java]: done

File: dependencies/src/main/java/com/google/cloud/tools/opensource/dependencies/ArtifactProblem.java [hunk 13]
```diff
@@ -35,6 +35,8 @@
 
   protected final ImmutableList<DependencyNode> dependencyPath;
 
+  // todo since the artifact is the last element in the list, there's really no
+  // reason to pass the artifact here. 
   protected ArtifactProblem(Artifact artifact, List<DependencyNode> dependencyPath) {
```
Comment (suztomo) [inline dependencies/src/main/java/com/google/cloud/tools/opensource/dependencies/ArtifactProblem.java]: This non-public constructor allows artifact problems in which we don’t know dependencyPath.
